src/torchmetrics/functional/nominal/thiels_u.py

```diff
@@ -0,0 +1,155 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+import warnings
+from collections import Counter
+from typing import Optional, Union
+
+import scipy.stats as ss
+import torch
+from torch import Tensor
+from typing_extensions import Literal
+
+from torchmetrics.functional.nominal.utils import _handle_nan_in_data
+
+
+def _thiels_u_input_validation(nan_strategy: str, nan_replace_value: Optional[Union[int, float]]) -> None:
+    if nan_strategy not in ["replace", "drop"]:
+        raise ValueError(
+            f"Argument `nan_strategy` is expected to be one of `['replace', 'drop']`, but got {nan_strategy}"
+        )
+    if nan_strategy == "replace" and not isinstance(nan_replace_value, (int, float)):
+        raise ValueError(
+            "Argument `nan_replace` is expected to be of a type `int` or `float` when `nan_strategy = 'replace`, "
+            f"but got {nan_replace_value}"
+        )
+
+
+def conditional_entropy(
+    preds: Tensor,
+    target: Tensor,
+    log_base: float = math.e,
+    nan_strategy: Literal["replace", "drop"] = "replace",
+    nan_replace_value: Optional[Union[int, float]] = 0.0,
+) -> Tensor:
+    r"""Computes conditional entropy of x given y: S(x|y) The implementation is based on the code provided in
+    https://github.com/shakedzy/dython/
+
+    Args:
+        preds: 1D or 2D tensor of categorical (nominal) data
+            - 1D shape: (batch_size,)
+            - 2D shape: (batch_size, num_classes)
+        target: 1D or 2D tensor of categorical (nominal) data
+            - 1D shape: (batch_size,)
+            - 2D shape: (batch_size, num_classes)
+        log_base: float value to be used as the base for logarithm
+        nan_strategy: Indication of whether to replace or drop ``NaN`` values
+        nan_replace_value: Value to replace ``NaN``s when ``nan_strategy = 'replace'``
+
+    Returns:
+        S(x|y): float
+
+    Example:
+        >>> from torchmetrics.functional import conditional_entropy
+        >>> _ = torch.manual_seed(42)
+        >>> preds = torch.randint(10, (10,))
+        >>> target = torch.randint(10, (10,))
+        >>> conditional_entropy(preds, target)
+        tensor(0.2773)
+    """
+    # reduce dim if ndim == 2
+    preds = preds.argmax(1) if preds.ndim == 2 else preds
+    target = target.argmax(1) if target.ndim == 2 else target
+    # handle the nan in data
+    preds, target = _handle_nan_in_data(preds, target, nan_strategy, nan_replace_value)
+
+    # set x and y
+    x, y = preds.numpy(), target.numpy()
+
+    y_counter = Counter(y)
+    xy_counter = Counter(list(zip(x, y)))
+    total_occurrences = sum(y_counter.values())
+    entropy = 0.0
+    for xy in xy_counter.keys():
+        p_xy = xy_counter[xy] / total_occurrences
+        p_y = y_counter[xy[1]] / total_occurrences
+        entropy += p_xy * math.log(p_y / p_xy, log_base)
+    return torch.tensor(entropy)
```

Comment:
Would it be possible this part vectorize with `torch`? Not sure at the first glance, but maybe it's doable 🤔

Reply:
Fixed it using the confmat based solution used in crammer's v implementation.